lib/phoenix/controller/flash.ex

```diff
@@ -0,0 +1,80 @@
+defmodule Phoenix.Controller.Flash do
+  import Plug.Conn
+
+  @moduledoc """
+  Helper module to fetch the flash message.
+
+  This helper will look at the session for a "phoenix_flash"
+
+  ## Examples
+
+      plug :fetch_flash
+  """
+
+  @session_key "phoenix_flash"
+  @session_atom :phoenix_flash
+
+  @doc """
+  Fetches the flash from server, or a signed message from the client
+
+  We will look for a flash message in the `session_key`, and if none
+  found then we will look in client cookie (under the name
+  `__session_key___`)
+  """
+  def fetch_flash(conn, _opts \\ []) do
+    found_flash = get_session(conn, @session_key)
+    conn = persist_flash(conn, found_flash || %{})
+
+    register_before_send conn, fn conn ->
+      flash = conn.private[@session_atom]
```

Comment:
@josevalim the original code had `conn.private.phoenix_flash`; I changed to use a constant up above.  This will no longer throw an exception if not found; not sure if that tweak is desired or not.

Reply:
Undesired. I would prefer to raise as it would mean something definitely went wrong. :)

Reply:
I added `|| raise KeyError, key: @session_atom, term: conn.private` so that we could keep the constants

Reply:
I tested locally with

```
iex(1)> atom = :c
:c
iex(2)> map = %{b: 1}
%{b: 1}
iex(3)> map[atom] || raise KeyError, key: atom, term: map
** (KeyError) key :c not found in: %{b: 1}
```

Reply:
Then it is best to use `Map.fetch!(conn.private, @session_atom)`, same result, more concise.

Reply:
It feels like you know this language inside and out 😜